docs/contributing/plugin-tutorial.md

```diff
@@ -29,254 +64,153 @@ Here is a visual statement of what we want to accomplish:
-                }
-```
+-- hls-plugin-api/src/Ide/Types.hs
 
-The meat is in the `commandFunc` field, which is of type `CommandFunction`, another type synonym from `LSP.Types`:
-```haskell
-type CommandFunction a =
-  LSP.LspFuncs Config
-  -> IdeState
+type CommandFunction ideState a
+  = ideState
   -> a
-  -> IO (Either ResponseError Value, Maybe (ServerMethod, ApplyWorkspaceEditParams))
+  -> LspM Config (Either ResponseError Value)
 ```
 
-`CommandFunction` takes in the familiar `LspFuncs` and `IdeState` arguments, together with a JSON encoded argument.
-I recommend checking the LSP specifications in order to understand how commands work, but briefly the LSP server (us) initially sends a command descriptor to the client, in this case as part of a code lens. When the client decides to execute the command on behalf of a user action (in this case a click on the code lens), the client sends this descriptor back to the LSP server which then proceeds to handle and execute the command. The latter part is implemented by the `commandFunc` field of our `PluginCommand` value.
 
-For our command, we are going to have a very simple handler that receives a diff (`WorkspaceEdit`) and returns it to the client. The diff will be generated by our code lens provider and sent as part
-of the code lens to the LSP client, who will send it back to our command handler when the user activates
-the code lens:
-```haskell
-importCommandId :: CommandId
-importCommandId = "ImportLensCommand"
+`CommandFunction` takes an `ideState` and a JSON-encodable argument.
 
-importLensCommand :: PluginCommand
-importLensCommand =
-    PluginCommand importCommandId "Explicit import command" runImportCommand
+Our handler will ignore the state argument and only use the `WorkspaceEdit` argument.
 
+```haskell
 -- | The type of the parameters accepted by our command
-data ImportCommandParams = ImportCommandParams WorkspaceEdit
-  deriving Generic
+newtype ImportCommandParams = ImportCommandParams WorkspaceEdit
+  deriving (Generic)
   deriving anyclass (FromJSON, ToJSON)
 
 -- | The actual command handler
-runImportCommand :: CommandFunction ImportCommandParams
-runImportCommand _lspFuncs _state (ImportCommandParams edit) = do
-    return (Right Null, Just (WorkspaceApplyEdit, ApplyWorkspaceEditParams edit))
-
+runImportCommand :: CommandFunction IdeState ImportCommandParams
+runImportCommand _ (ImportCommandParams edit) = do
+  -- This command simply triggers a workspace edit!
+  _ <- sendRequest SWorkspaceApplyEdit (ApplyWorkspaceEditParams Nothing edit) (\_ -> pure ())
+  return (Right Null)
 ```
 
+`runImportCommand` [sends a request](https://hackage.haskell.org/package/lsp/docs/Language-LSP-Server.html#v:sendRequest) to the client using the method `SWorkspaceApplyEdit` and the parameters `ApplyWorkspaceEditParams Nothing edit`, providing a response handler that does nothing. It then returns `Right Null`, which is an empty `Aeson.Value` wrapped in `Right`.
```

Comment:
Maybe explain why we have an `Either` type here, and that `Left` is used for failures?

Reply:
I have moved the description of the Either up to when we introduce `CommandFunction`. Please have a look.